src/libsyntax/parse/parser/expr.rs

```diff
@@ -1448,7 +1450,7 @@ impl<'a> Parser<'a> {
 
         Ok(ast::Arm {
             attrs,
-            pats,
+            pats: pat, // FIXME(or_patterns, Centril | dlrobertson): this should just be `pat,`.
```

Comment:
Why would this just be a singular `pat`? `Foo(A | B)` will now be a single pattern that contains an or-pattern, but woudn't we still want to parse `Foo(A) | Bar(B)` as multiple patterns?

Reply:
Having `arm.pat: P<Pat>` instead of `arm.pats: Vec<P<Pat>>` simplifies the overall structure of the compiler by using a uniform mechanism for or-patterns instead of having an distinction between or-patterns at the top-level and in nested positions. As an example, you can see that the parser `pat.rs` parser has, in this PR, a hack to unpack an or-pattern at the top level. There are many other places (e.g. in resolve, save-analysis, etc.) in which it would simplify to have one uniform mechanism. So we should represent `p1 | p2` at the top of a match arm simply as `PatKind::Or`.

Reply:
Got it! This makes a lot of sense and I'd love to see this happen.The lowering in MIR will require quite a bit of work and thought in order to get to this point, but AFAIK unifying the two could also greatly simplify `librustc_mir/build/matches`.

Reply:
Oh absolutely. I think basically we want to work from both sides and meet in the middle. That is, we fix the AST and add a hack to lowering. Then we fix lowering + HIR and add a hack to HAIR. Eventually we get down to MIR and the hack is no more.